examples/nlp/language_modeling/prepare_truncated_data.py

```diff
@@ -54,27 +59,215 @@ def _process_line(line, task, tokenizer, max_seq_length, tokens_to_generate_extr
+    elif truncation_pos == 'left':
+        if total_len > max_seq_length:
+            truncation_seperator = "[The beginning of the transcript is omitted] ..."
+
+            # Split the context into two parts at the last occurrence of the context string
+            # Truncate the suffix from the right side to ensure that its length is equal to the maximum allowed length for the truncated text
+            context = context_text
+            truncations_tokens = tokenizer.text_to_tokens(truncation_seperator)
+            truncated_context = context[
+                -(
+                    max_seq_length
+                    - tokens_to_generate_extra
+                    - len(tokenized_inputs)
+                    - len(tokenized_prompt)
+                    - len(truncations_tokens)
+                ) :
+            ]
+            truncated_text = TASKS[task]['template'].format(
+                context=truncation_seperator + truncated_context, input=input_text
+            )
+        else:
+            truncated_text = TASKS[task]['template'].format(context=context_text, input=input_text)
+    elif truncation_pos == 'middle':
+        if total_len > max_seq_length:
+            truncation_seperator = "... [The middle of the transcript is omitted] ..."
+            # Calculate the number of tokens to truncate from each side
+            truncations_tokens = tokenizer.text_to_tokens(truncation_seperator)
+            num_tokens_to_truncate = (
+                total_len
+                - max_seq_length
+                + tokens_to_generate_extra
+                + len(tokenized_inputs)
+                + len(tokenized_prompt)
+                + len(truncations_tokens)
+            )
+            num_tokens_to_truncate_half = num_tokens_to_truncate // 2
+
+            # Find the middle of context_tokens and truncate num_tokens_to_truncate tokens from the middle
+            middle = len(tokenized_contexts) // 2
+            prefix_tokens = tokenized_contexts[: middle - num_tokens_to_truncate_half]
+            suffix_tokens = tokenized_contexts[middle + num_tokens_to_truncate_half :]
+
+            # Convert the truncated parts back to text and concatenate them with the truncation separator and the question to create the final truncated text
+            truncated_prefix = tokenizer.tokens_to_text(prefix_tokens)
+            truncated_suffix = tokenizer.tokens_to_text(suffix_tokens)
+            truncated_context = truncated_prefix + truncation_seperator + truncated_suffix
+            truncated_text = TASKS[task]['template'].format(context=truncated_context, input=input_text)
+            # print(len(tokenizer.text_to_tokens(line['context'])),len(tokenizer.text_to_tokens(truncated_context)))
+        else:
+            truncated_text = TASKS[task]['template'].format(context=context_text, input=input_text)
+            # print(len(tokenizer.text_to_tokens(truncated_text)))
```

Comment:
Since this logic seems to be common for both zero scrolls and long bench, can we make this a separate function?

Reply:
@paul-gibbons

Reply:
I see. Okay.